torch/utils/_benchmark/utils/valgrind_wrapper.py

```diff
@@ -0,0 +1,227 @@
+
+        # We include `pass` so it matches the baseline and reduces the instruction diff.
+        fn_counts = self._invoke(
+            stmt=f"pass;{stmt}", setup=setup, number=number, num_threads=num_threads)
+
+        count_dict = {fn: count for count, fn in fn_counts}
+        for count, fn in self._baseline_cache[(number, num_threads)]:
+            count_dict.setdefault(fn, 0)
+            count_dict[fn] -= count
+            if not count_dict[fn]:
+                count_dict.pop(fn)
+
+        fn_diff_counts = []
+        for fn, count in sorted(count_dict.items(), key=lambda x: x[1], reverse=True):
+            if count < 0:
+                # The profiling process is not 100% deterministic. For instance
+                # the subprocess calls to control callgrind poll resulting in
+                # a non-deterministic number of instructions. This delta is
+                # generally small.
+                # FIXME: Validate that differences are small and expected, and warn otherwise.
+                break
+            fn_diff_counts.append((count, fn))
+        return fn_diff_counts
+
+    def _invoke(self, stmt: str, setup: str, number: int, num_threads: int):
+        working_dir = tempfile.mkdtemp()
+
+        _, script_file = tempfile.mkstemp(prefix="timer_callgrind_", suffix=".py", dir=working_dir)
+        _, callgrind_out = tempfile.mkstemp(prefix="callgrind.", suffix=".out", dir=working_dir)
+        _, error_log = tempfile.mkstemp(prefix="error_", suffix=".txt", dir=working_dir)
+
+        try:
+            with open(script_file, "wt") as f:
+                f.write(self._construct_script(
+                    stmt=stmt, setup=setup, number=number,
+                    num_threads=num_threads, error_log=error_log))
+
+            valgrind_invocation = subprocess.run(
+                [
+                    "valgrind",
+                    "--tool=callgrind",
+                    f"--callgrind-out-file={callgrind_out}",
+                    "--dump-line=yes",
+                    "--dump-instr=yes",
+                    "--collect-jumps=yes",
+                    "--instr-atstart=no",
+                    "python",
+                    script_file,
+                ],
+                capture_output=True
+            )
+            if valgrind_invocation.returncode:
+                with open(error_log, "rt") as f:
+                    error_report = f.read()
+                if not error_report:
+                    error_report = "Unknown error."
+                    error_report += "\n" + valgrind_invocation.stdout.decode("utf-8")
+                    error_report += "\n" + valgrind_invocation.stderr.decode("utf-8")
+                raise OSError(f"Failed to collect callgrind profile:\n{error_report}")
+
+            annotate_invocation = subprocess.run(
+                [
+                    "callgrind_annotate",
+                    "--inclusive=no",
```

Comment:
Having this to be "no" is fine - just FYI, for some analysis like RecordFunction's impact on autograd, having this set to yes can also indicate whether the __tls_get_addr increase is mostly from RecordFunction or not (RecordFunction itself costs ~48 insts, but inclusive costs ~201 insts, mostly from TLS).

Reply:
I had mostly done that because you need exclusive counts to do meaningful diffs in a lot of cases. But the truth is the endpoint really shouldn't be making such strong assumptions about how a caller wants to use the measurement, so instead I packaged both the inclusive and exclusive stats into a `CallgrindStats` object to allow users to choose how they want to slice the stats.